### ISSUE TITLE
CarbonFootPrint as a mobile app

### ISSUE THREAD

**Issue by boro741** [OiWorld/CarbonFootprint#135]
CarbonFootPrint as a mobile app would help people to get awareness how much amount of Carbon they are emitting while travelling.

**Comment by ceilican** [comment]
It would be nice to use the user's location.

I imagine that we could try to figure out (or ask the user) whether he/she is moving by car, bicycle or public transportation and poll his location every few seconds and use the distance travelled to estimate the CO2 emissions. We could show update the CO2 emissions in real time! That could be very impactful. Such an app would be independent of (and different from) the browser extensions.

**Comment by Kolpa** [comment]
https://developers.google.com/android/reference/com/google/android/gms/location/ActivityRecognitionApi

This could be quite relevant

**Comment by ceilican** [comment]
Interesting, @Kolpa !

**Comment by boro741** [comment]
I'm thinking of building with nativescript  https://www.nativescript.org/

**Comment by ceilican** [comment]
Nativescript seems to be a good option. Are there any competing alternatives?

**Comment by boro741** [comment]
yeah. React native http://facebook.github.io/react-native/docs/getting-started.html .

**Comment by ceilican** [comment]
What are the advantages and disadvantages of each?

**Comment by boro741** [comment]
Both supports the native API's.
What I think the advantage of Native Script is that we can easily build with JavaScript.
For React Native it requires knowledge of react.js.

Or,
We can go with java but it wouldn't be supporting on cross platform.

**Comment by heychirag** [comment]
Hi @ceilican,

The app would be great addition to the project!

> It would be nice to use the user's location.

I have a slightly different approach to how we may be able to accomplish this. The official Google Maps  app from Google tracks down the entire location data of every person and shows a summary of the person's day in the "Your Timeline Pane". I think we may be able to retrieve this data through their APIs. The best part of this would be that we wouldn't have to worry about whether the user is walking, driving or using public transport since we would be relying on Google's efficient algorithms.

![screenshot_20170120-230638](https://cloud.githubusercontent.com/assets/13611666/22159762/7245bb2a-df68-11e6-836c-690db0ee3b5c.png)

Let me know what do you think about this.

**Comment by ceilican** [comment]
So, it seems that we have three ways of getting the information we need:

1) extracting the information we need from Google Maps app (assuming that it could be done somehow), as suggested by @heychirag .

2) using the Activity Recognition API suggested by @Kolpa .

3) polling location data ourselves and doing the activity recognition and time and distance measurements ourselves.


Regarding option 1: can it be done? Does Google Maps API really make all that information available to other apps?

Regarding option 2: it is clear that the API recognizes the current activity for us. But does it also keep track of time and distance spent on each activity? Or would we have to do that ourselves?

Which of these two solutions is more likely to require less maintenance in the long term?

**Comment by Kolpa** [comment]
Regarding option one: this would be great but as far as I know Google does not allow qfdws

**Comment by ceilican** [comment]
what is "qfdws"?

**Comment by Kolpa** [comment]
That was not intended sorry about that I edited the original comment

**Comment by heychirag** [comment]
Although there is no official API, the data can be scrapped from here  

https://www.google.com/maps/timeline/kml

If, you have allowed Google to access your location on your mobile, the history can be found in the above link.

**Option 1:**

Pros
- Straightforward - Using a feature that's already available
- Easy cross platform production/maintenance

Cons
-  Not everyone uses Google Maps

**Option 2:**

Pros
- Independent - Doesn't rely on others

Cons
- Processing of raw GPS data may be challenging 
- Difficult cross platform production/maintenance

**Comment by boro741** [comment]
So the conclusion would be building a native app like Uber app using JavaScript frameworks(native script, react native) so that we don't need to implement everything from scratch.

Also use of js frameworks like native script would solve the cross-platform issue although it might be bit slower than native apps which are built on java or swift.

**Comment by ceilican** [comment]
From now on, discussion of this issue should continue on [this issue](https://www.gitlab.com/aossie/CarbonFootprint/issues/137) in our new issue tracker in GitLab,

**Comment by ajayrahul97** [comment]
Hi , I am Ajay Rahul , a sophomore at IIT Roorkee . I am an Android Developer , and i would like to contribute to the Carbon FootPrint for Android project as a part of GSoC 17 . Could you please comment the link for the github repo for the same ?, also , I believe carbon footprint for mobile would create more impact and be more efficient in achieving the objective !

**Comment by heychirag** [comment]
@ajayrahul97 There is no project repo yet. If you want to work on it, I'll be glad to setup one for you.

**Comment by ajayrahul97** [comment]
@heychirag please do so , i am willing to work on it , i will have a look at the chrome extension and try to replicate it and also improvise it as a part of my GSoC 17 program ! Thanks !

**Comment by DravitLochan** [comment]
Hi! My name is Dravit, a 2nd year undergrad pursuing Computer Science Engineering from Jaypee Institute of Information Technology. It is one cool and interesting idea. I look forward to contribute to Carbon Footprint in GSoC'17.

**Comment by ceilican** [comment]
Hi @DravitLochan and @ajayrahul97 ! Remember that we moved to http://www.gitlab.com/aossie/CarbonFootprint/ .

**Comment by ceilican** [comment]
See also the instruction in http://www.aossie.org/ .

**Comment by shikharbhuddi** [comment]
Hi everyone! Myself Shikhar, undergraduate from BITS Pilani. I find the idea of Carbon Footprint intriguing and I am willing to work on it, with fellow developers as a part of GSoC 2017. I have prior experience in full Android application development lifecycle. I have worked with open source maps libraries as well.

**Comment by Chaitanyachavali** [comment]
Hello everyone, I am Chaitanya, 3rd year undergraduate from NIIT University. Porting CarbonFootprint from browser extension to mobile application is a great idea. I want to work in this development as a part of GSoC 2017.
Proposal: Why can't we develop the mobile application in Apache Cordova (https://cordova.apache.org/). Cordova framework helps in building  Android,IOS,Windows applications with a single codebase. It majorly involves with Javascript. I have gone through the source of CarbonFootprint (chrome's source) and figured out that it is possible to develop a mobile application for three platforms in Cordova. If you find this solution acceptable, help me in proceeding to the further steps.
**Note**: I have already developed mobile applications in Cordova and used google's API (charts, OAuth). Check my repositories to find them. I have already developed and built (android by now) a simple mobile application which can take the geolocation of the user. Source can be shared if needed.

**Comment by ceilican** [comment]
@Chaitanyachavali : Thank you for your ideas! Could you repost them in our gitlab issue tracker (https://gitlab.com/aossie/CarbonFootprint/issues/137). We are gradually moving from Github to Gitlab.

**Comment by ceilican** [comment]
@shikharbhuddi : thank for the intro. However, note that you shouldn't use issue trackers for personal introductions. Please have a look on our website (http://www.aossie.org) to find out how to contribute to our projects and communicate with us.

**Comment by harshitdiwedi03** [comment]
Hello Everyone!
I am a sophomore at IIT Delhi and wan to work on this project. I tries searching but couldn't find any repo for it. Can someone please guide me regarding how to start contributing for this project?

**Comment by cskc97** [comment]
Hi! I'm Krish, a sophomore at NYU. Came across this project and thought it was really cool. I'm gonna apply through the GSOC program. Can anyone let me know how I can join the mailing lists and also as to how I can contribute to the project? Thanks!

**Comment by ceilican** [comment]
@cskc97 Hi! All the instructions that you need can be found in www.aossie.org .

**Comment by argon448** [comment]
I am a sophomore at AKTU and eager to contribute to this lovely project through GSoC.
